gen3utils/etl/etl_validator.py

```diff
@@ -1,202 +1,274 @@
+            )
+        )
+    return name
+
+
+def validate_name_src(json_obj, path, recorded_errors, nodes_with_props):
+    name = validate_name(json_obj, recorded_errors)
+    fn = validate_fn(json_obj, recorded_errors)
+    src = json_obj.get("src", name)
+    if src is not None:
+        if path is None:
+            recorded_errors.append(
+                FieldError(
+                    "src field must be specified with a path for {}".format(json_obj)
+                )
+            )
+        else:
+            path_items = path.split(".")
+            if fn != "count" and src not in nodes_with_props[path_items[-1]]:
+                recorded_errors.append(
+                    FieldError(
+                        "src field {} (declared in {}) is not found in given dictionary.".format(
+                            src, json_obj
+                        )
+                    )
+                )
+    return name
+
+
+def validate_path(
+    json_obj, grouping_path, recorded_errors, list_of_nodes, nodes_with_props
+):
+    path = json_obj.get("path", grouping_path)
+    names = []
+    if path is None:
+        recorded_errors.append(
+            PropertiesError(
+                "Missing path declaration for the property {}.".format(json_obj)
+            )
+        )
+    else:
+        path_items = path.split(".")
+        if "_ANY" in path_items:
+            path_items.remove("_ANY")
+        for item in path_items:
+            [edge, str_fields] = (
+                list(filter(None, re.split(r"[\[\]]", item)))
+                if item.find("[") != -1
+                else [item, None]
+            )
+            if edge not in list_of_nodes:
+                recorded_errors.append(PathError(path))
+            if str_fields is not None:
+                fields = str_fields.split(",")
+                for f in fields:
+                    src = f.split(":")[-1] if f.find(":") != -1 else f
+                    name = f.split(":")[0] if f.find(":") != -1 else f
+                    names.append(
+                        validate_name_src(
+                            {"name": name, "src": src},
+                            edge,
+                            recorded_errors,
+                            nodes_with_props,
+                        )
+                    )
+    return names
+
+
+def get_all_nodes(model):
+    labels_to_back_refs = {}
+    nodes_with_props = {}
+    categories_to_labels = {}
+    all_classes = model.Node.get_subclasses()
+    for n in all_classes:
+        present_node = n._pg_edges
+        present_props = n.__pg_properties__
+        present_node_props = [k for k in present_props.keys()]
+        present_node_props.append("id")
+        category = n._dictionary.get("category")
+        if category not in categories_to_labels:
+            categories_to_labels[category] = []
+        categories_to_labels.get(category).append(n.label)
```

Comment:
what do you mean using defaultdict?

Reply:
to avoid checking if the key already exists or not, you can use defaultdict
```
from collections import defaultdict
categories_to_labels = defaultdict(list)
for [...]:
    categories_to_labels[category].append(n.label)
```

Reply:
thanks